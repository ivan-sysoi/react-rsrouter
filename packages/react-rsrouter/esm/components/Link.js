import * as React from 'react';
import * as invariant from 'invariant';
import { RouterProviderContext } from './contexts';
import history from '../history';
const isModifiedEvent = (event) => Boolean(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);
class Link extends React.PureComponent {
    constructor() {
        super(...arguments);
        this.state = {};
        this.onClick = (event) => {
            if (!event.defaultPrevented && // onClick prevented default
                event.button === 0 && // ignore everything but left clicks
                !this.props.target && // let browser handle "target=_blank" etc.
                !isModifiedEvent(event) // ignore clicks with modifier keys
            ) {
                event.preventDefault();
                if (history) {
                    if (Boolean(this.props.replace)) {
                        history.replace(this.state.href);
                    }
                    else {
                        history.push(this.state.href);
                    }
                }
            }
        };
    }
    static getDerivedStateFromProps(nextProps, prevState) {
        invariant(nextProps.to, 'You must specify the "to" property');
        if (prevState.propsTo !== nextProps.to || prevState.propsParams !== nextProps.params) {
            const to = Array.isArray(nextProps.to) ? nextProps.to : nextProps.to.split('.');
            const params = nextProps.params || {};
            return {
                to,
                params,
                propsTo: nextProps.to,
                propsParams: nextProps.params,
                href: nextProps.buildUrl(to, params),
            };
        }
        return null;
    }
    render() {
        return (React.createElement("a", { href: this.state.href, onClick: this.onClick }, this.props.children));
    }
}
Link.defaultProps = {
    replace: false,
};
const LinkWithBuildUrl = props => {
    return (React.createElement(RouterProviderContext.Consumer, null, ({ buildUrl }) => {
        return React.createElement(Link, Object.assign({}, props, { buildUrl: buildUrl }));
    }));
};
export default LinkWithBuildUrl;
//# sourceMappingURL=Link.js.map