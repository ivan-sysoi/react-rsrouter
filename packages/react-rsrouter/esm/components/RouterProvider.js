import * as React from 'react';
import { connect } from 'react-redux';
import * as invariant from 'invariant';
import { RouterProviderContext, RouteContext } from './contexts';
import { selectMatch } from '../store';
export class RouterProvider extends React.Component {
    constructor() {
        super(...arguments);
        this.buildUrl = (to, params) => {
            return this.props.routes.buildUrl(to, params, this.props.getState);
        };
    }
    render() {
        if (this.props.match) {
            return (React.createElement(RouterProviderContext.Provider, { value: { buildUrl: this.buildUrl } },
                React.createElement(RouteContext.Provider, { value: { path: this.props.match.path } }, this.props.children)));
        }
        return null;
    }
}
const stateToProps = (state) => {
    invariant(state.router, 'Router state is not found. You must attach it under "router" key.');
    return {
        match: selectMatch(state),
        getState: () => state,
    };
};
export const ConnectedRouterProvider = connect(stateToProps)(RouterProvider);
export default ConnectedRouterProvider;
//# sourceMappingURL=RouterProvider.js.map