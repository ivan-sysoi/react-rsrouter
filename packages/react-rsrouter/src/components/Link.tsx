import * as React from 'react'
import * as invariant from 'invariant'

import { RouterProviderContextValues, RouterProviderContext } from './contexts'
import history from '../history'
import { RouteParams, RoutePath } from '../'

type propsTo = RoutePath | string

interface LinkProps {
  to: propsTo
  params?: RouteParams
  target?: string
  replace?: boolean
}

interface LinkPropsWithBuildUrl extends LinkProps {
  buildUrl: (to: RoutePath, params: RouteParams) => string
}

interface LinkState {
  to: RoutePath
  params: RouteParams
  href: string
  propsTo: propsTo
  propsParams?: RouteParams
}

const isModifiedEvent = (event: React.MouseEvent<HTMLAnchorElement>): boolean =>
  Boolean(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey)

class Link extends React.PureComponent<LinkPropsWithBuildUrl, LinkState> {
  static defaultProps = {
    replace: false,
  }

  static getDerivedStateFromProps(nextProps: LinkPropsWithBuildUrl, prevState: LinkState): LinkState | null {
    invariant(nextProps.to, 'You must specify the "to" property')

    if (prevState.propsTo !== nextProps.to || prevState.propsParams !== nextProps.params) {
      const to = Array.isArray(nextProps.to) ? nextProps.to : nextProps.to.split('.')
      const params = nextProps.params || {}
      return {
        to,
        params,
        propsTo: nextProps.to,
        propsParams: nextProps.params,
        href: nextProps.buildUrl(to, params),
      }
    }

    return null
  }

  readonly state: LinkState = {} as LinkState

  onClick = (event: React.MouseEvent<HTMLAnchorElement>): void => {
    if (
      !event.defaultPrevented && // onClick prevented default
      event.button === 0 && // ignore everything but left clicks
      !this.props.target && // let browser handle "target=_blank" etc.
      !isModifiedEvent(event) // ignore clicks with modifier keys
    ) {
      event.preventDefault()
      if (history) {
        if (Boolean(this.props.replace)) {
          history.replace(this.state.href)
        } else {
          history.push(this.state.href)
        }
      }
    }
  }

  render() {
    return (
      <a href={this.state.href} onClick={this.onClick}>
        {this.props.children}
      </a>
    )
  }
}

const LinkWithBuildUrl: React.SFC<LinkProps> = props => {
  return (
    <RouterProviderContext.Consumer>
      {({ buildUrl }: RouterProviderContextValues) => {
        return <Link {...props} buildUrl={buildUrl} />
      }}
    </RouterProviderContext.Consumer>
  )
}

export default LinkWithBuildUrl
