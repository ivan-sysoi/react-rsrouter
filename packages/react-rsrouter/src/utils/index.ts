import { RouterLocation } from '..'

export * from './routes'
export * from './matcher'

export const isDiffLocations = (loc1: RouterLocation, loc2: RouterLocation): boolean =>
  loc1.pathname !== loc2.pathname || loc1.search !== loc2.search

export const locationToUrl = (location: RouterLocation): string =>
  `${location.pathname}${location.search ? `?${location.search}` : ''}`

export const urlToRouterLocation = (url: string): RouterLocation => {
  const [pathname, search] = url.split('?')
  return { pathname, search }
}
