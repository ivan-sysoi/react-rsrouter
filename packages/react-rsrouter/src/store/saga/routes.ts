import { Saga, Task, SagaIterator } from 'redux-saga'
import { fork, cancel, put } from 'redux-saga/effects'

import { IRoutesCollection } from '../../utils'
import { setMatch } from '../actions'
import { Match } from '../../.'

export type applyMatchSaga = ({ match }: { match: Match }) => SagaIterator

export const createApplyMatch = ({ routes }: { routes: IRoutesCollection }): applyMatchSaga => {
  const runTasks: Task[] = []
  return function* applyMatch({ match }: { match: Match }) {
    while (runTasks.length > 0) {
      yield cancel(runTasks.pop() as Task)
    }
    yield put(setMatch(match))
    for (const route of routes.getRoutes(match.path)) {
      if ('onEnter' in route) {
        const onEnter = route.onEnter as Saga
        runTasks.push(yield fork(onEnter, { ssr: false }))
      }
    }
  }
}
