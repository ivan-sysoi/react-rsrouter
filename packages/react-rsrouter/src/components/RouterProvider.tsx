import * as React from 'react'
import { connect } from 'react-redux'
import * as invariant from 'invariant'

import { RouterProviderContext, RouteContext } from './contexts'
import { stateGetter, RootState, selectMatch } from '../store'
import { IRoutesCollection } from '../utils/routes'
import { RouteParams, RoutePath, Match } from '../.'

export interface RouterProviderProps {
  routes: IRoutesCollection
  match: Match
  getState: stateGetter
}

export class RouterProvider extends React.Component<RouterProviderProps> {
  buildUrl = (to: RoutePath, params: RouteParams): string => {
    return this.props.routes.buildUrl(to, params, this.props.getState)
  }

  render() {
    if (this.props.match) {
      return (
        <RouterProviderContext.Provider value={{ buildUrl: this.buildUrl }}>
          <RouteContext.Provider value={{ path: this.props.match.path }}>{this.props.children}</RouteContext.Provider>
        </RouterProviderContext.Provider>
      )
    }
    return null
  }
}

const stateToProps = (state: RootState) => {
  invariant(state.router, 'Router state is not found. You must attach it under "router" key.')
  return {
    match: selectMatch(state),
    getState: () => state,
  }
}
export const ConnectedRouterProvider = connect(stateToProps)(RouterProvider)

export default ConnectedRouterProvider
