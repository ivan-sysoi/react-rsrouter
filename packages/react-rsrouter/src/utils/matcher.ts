import { Match, RouteParams, RouterLocation } from '..'
import { RouteSchema } from './routes'

export interface MatcherContext {
  getState: () => object
  match: Match
}

export type LocationFullMatch = { params?: RouteParams; nextLocation: null }
export type LocationPartResult = { params?: RouteParams; nextLocation: RouterLocation }
export type LocationNoMatch = null
export type MatcherResult = LocationFullMatch | LocationPartResult | LocationNoMatch

export function matchLocation(route: RouteSchema, location: RouterLocation, context: MatcherContext): MatcherResult {
  switch (route.type) {
    case 'static': {
      const routeExpectsExactMatch = route.path.endsWith('$')
      const routePath = routeExpectsExactMatch ? route.path.slice(0, -1) : route.path

      if (routeExpectsExactMatch && routePath === location.pathname) {
        return {
          nextLocation: null,
        }
      }
      if (location.pathname.startsWith(routePath)) {
        const hasNested = Array.isArray(route.nested) && route.nested.length > 0

        if (!routeExpectsExactMatch || hasNested) {
          return {
            nextLocation: {
              ...location,
              pathname: location.pathname.slice(routePath.length),
            },
          }
        }
      }
      return null
    }
    case 'dynamic': {
      return route.pathMatcher(location, context)
    }
    case 'fallback': {
      return {
        nextLocation: null,
      }
    }
    default: {
      throw new Error(`Unknown route schema:\n${JSON.stringify(route, null, 4)}`)
    }
  }
}
