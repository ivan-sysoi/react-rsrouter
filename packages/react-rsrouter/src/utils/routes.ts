import { Saga } from 'redux-saga'
import * as jestValidate from 'jest-validate'

import { stateGetter } from '../store'
import { RouterLocation, RoutePath, RouteParams, Match } from '..'
import { MatcherResult, MatcherContext, matchLocation } from './matcher'

export interface BuildUrlContext {
  getState: stateGetter
}

export interface BaseRouteSchema {
  id: string
  nested?: RouteSchema[]
  onEnter?: Saga
}

export interface StaticRouteSchema extends BaseRouteSchema {
  type: 'static'
  path: string
}

export interface DynamicRouteSchema extends BaseRouteSchema {
  type: 'dynamic'
  pathMatcher: (location: RouterLocation, context: MatcherContext) => MatcherResult
  buildUrl: (params: RouteParams, context: BuildUrlContext) => string
}

export interface FallbackRouteSchema {
  id: string
  type: 'fallback'
}

export type RouteSchema = StaticRouteSchema | DynamicRouteSchema | FallbackRouteSchema

export interface IRoutesCollection {
  getRoutes(path: RoutePath): IterableIterator<RouteSchema>
  traverse(location: RouterLocation, context: MatcherContext): Match | null
  getMatch(location: RouterLocation, getState: stateGetter): Match
  getFallbackMatch(path?: RoutePath): Match
  isFallbackPath(path: RoutePath): boolean
  buildUrl(path: RoutePath, params: RouteParams, getState: stateGetter): string
}

function extendMatch(match: Match, routeId: string, params?: RouteParams): Match {
  return {
    path: [...match.path, routeId],
    params: {
      ...match.params,
      ...(params ? params : {}),
    },
  }
}

function extendMatcherContext(context: MatcherContext, routeId: string, params?: RouteParams): MatcherContext {
  return {
    ...context,
    match: extendMatch(context.match, routeId, params),
  }
}

function validateRouteSchema(route: RouteSchema) {
  let routeType
  if (route && 'type' in route) {
    routeType = route.type
  }

  switch (routeType) {
    case 'dynamic': {
      jestValidate.validate(route, {
        exampleConfig: {
          id: 'dynamic-route',
          type: 'dynamic',
          pathMatcher(location: RouterLocation, context: MatcherContext) {
            return { nextLocation: {}, params: {} }
          },
          buildUrl(path: RoutePath, params: RouteParams, context: BuildUrlContext) {
            return '/some/url'
          },
          nested: [],
          *onEnter() {
            yield
          },
        },
      })
      break
    }
    case 'static': {
      jestValidate.validate(route, {
        exampleConfig: {
          id: 'static-route',
          type: 'static',
          path: '/some/path$',
          nested: [],
          *onEnter() {
            yield
          },
        },
      })
      break
    }
    case 'fallback':
    default: {
      jestValidate.validate(route, {
        exampleConfig: {
          id: 'fallback-route',
          type: 'fallback',
          *onEnter() {
            yield
          },
        },
      })
      break
    }
  }
}

export type ParentRoutesCollection = { parentRouteId: string; collection: RoutesCollection }

export class RoutesCollection implements IRoutesCollection {
  protected readonly routes: Map<string, RouteSchema>
  protected readonly nested: Map<string, RoutesCollection>
  protected readonly parent: ParentRoutesCollection | null
  protected readonly fallbackId: string | null

  constructor(routes: RouteSchema[], parent?: ParentRoutesCollection) {
    this.fallbackId = null
    this.parent = parent || null
    this.routes = new Map()
    this.nested = new Map()
    for (const route of routes) {
      if (route.type === 'fallback') {
        this.fallbackId = route.id
      } else if (Array.isArray(route.nested) && route.nested.length > 0) {
        this.nested.set(route.id, new RoutesCollection(route.nested, { parentRouteId: route.id, collection: this }))
      }
      if (process.env.NODE_ENV !== 'production') {
        validateRouteSchema(route)
      }
      this.routes.set(route.id, Object.freeze(route))
    }
  }

  traverse(location: RouterLocation, context: MatcherContext): Match | null {
    for (const route of this.routes.values()) {
      const matcherResult = matchLocation(route, location, context)

      if (matcherResult !== null) {
        // is not exact match and has nested
        if (matcherResult.nextLocation !== null && this.nested.has(route.id)) {
          const nestedRoutes = this.nested.get(route.id) as RoutesCollection
          const res = nestedRoutes.traverse(
            matcherResult.nextLocation as RouterLocation,
            extendMatcherContext(context, route.id, matcherResult.params),
          )
          if (res !== null) {
            return res
          }
        } else {
          return extendMatch(context.match, route.id, matcherResult.params)
        }
      }
    }
    if (this.fallbackId !== null) {
      return extendMatch(context.match, this.fallbackId)
    }
    return null
  }

  getMatch(location: RouterLocation, getState: stateGetter): Match {
    const res = this.traverse(location, { getState, match: { path: [], params: {} } })
    if (res === null) {
      throw new Error('Route is not found. Did you forget to specify fallback route?')
    }
    return res
  }

  getRoutes(path: RoutePath): IterableIterator<RouteSchema> {
    let collection = this as RoutesCollection
    let ind = 0
    return {
      next() {
        if (ind < path.length) {
          if (!collection) {
            throw new Error(`Route ${path[ind]} is not found`)
          }
          const route = collection.routes.get(path[ind])
          if (!route) {
            throw new Error(`Route ${path[ind]} is not found`)
          }
          const res = { done: false, value: route as RouteSchema }
          collection = collection.nested.get(path[ind]) as RoutesCollection
          ind += 1
          return res
        }
        // @ts-ignore
        return { done: true, value: null as RouteSchema }
      },
      [Symbol.iterator]() {
        return this
      },
    }
  }

  private getRouteFullPath(routeId: string): RoutePath {
    let parent = this.parent
    const path = [routeId]
    while (parent) {
      path.push(parent.parentRouteId)
      parent = parent.collection.parent
    }
    path.reverse()
    return path
  }

  getFallbackMatch(path?: RoutePath): Match {
    if (Array.isArray(path) && path.length > 1 && this.nested.has(path[0])) {
      // go into deep to find the most nest collection
      let nestedCollection = this.nested.get(path[0]) as RoutesCollection
      for (let ind = 1; ind < path.length - 1; ind += 1) {
        if (!nestedCollection.nested.has(path[ind])) {
          break
        }
        nestedCollection = nestedCollection.nested.get(path[ind]) as RoutesCollection
      }

      // go upper, return the first fallback route
      while (nestedCollection) {
        if (nestedCollection.fallbackId !== null) {
          return {
            path: nestedCollection.getRouteFullPath(nestedCollection.fallbackId),
            params: {},
          }
        }
        if (nestedCollection.parent) {
          nestedCollection = nestedCollection.parent.collection
        } else {
          break
        }
      }
    }
    if (this.fallbackId === null) {
      throw new Error('You must specify at least one fallback route')
    }
    return { path: [this.fallbackId], params: {} }
  }

  isFallbackPath(path: RoutePath): boolean {
    const routes = Array.from(this.getRoutes(path))
    return routes[routes.length - 1].type === 'fallback'
  }

  buildUrl(path: RoutePath, params: RouteParams, getState: stateGetter): string {
    let url = ''
    const routeId = path[0]
    const route = this.routes.get(routeId) as RouteSchema
    if (route === undefined) {
      throw new Error(`Route ${routeId} is not found`)
    }

    switch (route.type) {
      case 'static': {
        url = route.path.endsWith('$') ? route.path.slice(0, -1) : route.path
        break
      }
      case 'dynamic': {
        url = route.buildUrl(params, { getState })
        break
      }
    }

    if (path.length > 1) {
      if (!this.nested.has(routeId)) {
        throw new Error(`Route ${routeId} hasn't nested routes`)
      }
      const nested = this.nested.get(routeId) as RoutesCollection
      return `${url}${nested.buildUrl(path.slice(1), params, getState)}`
    }
    return url
  }
}
