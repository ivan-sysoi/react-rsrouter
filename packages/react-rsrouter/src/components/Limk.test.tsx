import * as React from 'react'
import * as reactTestRenderer from 'react-test-renderer'
import Link from './Link'
import { RouterProvider } from './RouterProvider'
import { RoutesCollection } from '../utils/routes'

describe('Link', () => {
  it('should render url', () => {
    const state = {
      router: { match: { path: ['simple'], params: {} }, serverResponse: { status: 200 } },
    }
    const dummyGetState = () => state

    const routes = new RoutesCollection([
      {
        id: 'simple',
        type: 'static',
        path: '/some/path',
      },
    ])

    const link = reactTestRenderer.create(
      <RouterProvider routes={routes} match={state.router.match} getState={dummyGetState}>
        <Link to="simple">Simple Link</Link>
      </RouterProvider>,
    )

    expect(link.toJSON()).toMatchSnapshot()
  })
})
