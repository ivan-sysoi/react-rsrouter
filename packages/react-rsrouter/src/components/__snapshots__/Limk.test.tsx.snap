// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should render url 1`] = `
<a
  href="/some/path"
  onClick={[Function]}
>
  Simple Link
</a>
`;
